### DISCUSSION TITLE
`slam_toolbox` and LIDAR: error `discarding message because the queue is full`

### DISCUSSION THREAD

**Discussion by sebastianfer** [iRobotEducation/create3_docs#116] [Questions] [answered]
Hi, I am trying to generate a map using LIDAR, and I am following the instructions listed here: https://github.com/iRobotEducation/create3_examples/tree/e0dcd78f9b00c6fa1533d1d4cb37542c29095edc/create3_lidar 

My Raspberry Pi 4, robot, and virtual machine are all in the same network, and I am getting full topic list. I opened three terminals and logged into the Pi for the setup. The command that starts the sensors seems to be working fine (ros2 launch create3_lidar sensors_launch.py), but when I run slam toolbox in the second terminal (ros2 launch create3_lidar slam_toolbox_launch.py), I get the following message:**[async_slam_toolbox_node-1] [INFO] [1658262909.942375981] [slam_toolbox]: Message Filter dropping message: frame 'laser_frame' at time 1658262909.862 for reason 'discarding message because the queue is full'**

Any ideas how I can solve this?

**Comment by justinIRBT** [comment]
Is your Create3 on the network?  This error could be because your robot's system time doesn't match that of the computer your lidar is attached to.  If your robot is on the network, it will use NTP to sync to global time.  Alternatively, if you don't want to / can't get your Create3 on the network, you can sync the time between your SBC and Create3 following the instructions here:
https://iroboteducation.github.io/create3_docs/setup/compute-ntp/

**Reply by sebastianfer** [reply]
My Pi is connected to the Create3 by WiFi. I found out, however, that the Create3 is in a different time zone (UTC), while the Pi is in ET. Maybe that's causing the problem? If so, how could I solve it?

**Reply by justinIRBT** [reply]
If the Create is connected to the pi over wifi, then the NTP sync won't work as it specifies the USB interface, but if your Create is connected to wifi, it should be able to access a global ntp server unless your network is blocking it.  The time zone doesn't matter if the time is consistent between the time zones (ex create shows 6:16pm UTC and your pi shows 2:16pm in Boston).  To verify whether time is the problem, try doing a `ros2 topic echo` on the `odom` topic from the robot and separately your laser topic.  Look at the header timestamp in the printed message to see if they look similar (same value in the second field)

**Reply by justinIRBT** [reply]
What are you showing in the screenshots of the 2 topic prints?  They both look like prints from `/odom` at different times.  What I want to see is 2 terminals side by side, one echo'ing `/odom` and one echo'ing your laser topic (likely `/scan`).  Then we can see the timestamps side by side to see if there is a mismatch.

**Reply by sebastianfer** [reply]
Sorry for the confusion! Here are the screenshots of the Odom and the scan
https://drive.google.com/drive/folders/1RN4QhsM2V8MuWaUsttXvoj71xTPtpaVh?usp=sharing

**Reply by justinIRBT** [reply]
They seem to be about a second apart, but I am assuming that might just be from Ctrl+C the terminal at slightly different times, so time might not be your issue.  Can you post the full startup text for slam_toolbox until it starts repeating messages, you can just post it inline in the comments here

**Reply by sebastianfer** [reply]
[INFO] [launch]: All log files can be found below /home/pi/.ros/log/2022-07-20-16-17-33-278986-ubuntu-6917
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [async_slam_toolbox_node-1]: process started with pid [6919]
[async_slam_toolbox_node-1] [INFO] [1658348253.905401692] [slam_toolbox]: Node using stack size 40000000
[async_slam_toolbox_node-1] [INFO] [1658348253.994244786] [slam_toolbox]: Using solver plugin solver_plugins::CeresSolver
[async_slam_toolbox_node-1] [INFO] [1658348253.994963466] [slam_toolbox]: CeresSolver: Using SCHUR_JACOBI preconditioner.
[async_slam_toolbox_node-1] [INFO] [1658348254.181244707] [slam_toolbox]: Message Filter dropping message: frame 'laser_frame' at time 1658348254.101 for reason 'discarding message because the queue is full'

**Reply by justinIRBT** [reply]
Does the tf tree show up properly on `rviz2`?  Is you drive the robot and select the `odom` frame in rviz, is the laser frame moving with the robot if you display the tf tree?

**Reply by justinIRBT** [reply]
https://answers.ros.org/question/389383/slam_toolbox-message-filter-dropping-message-for-reason-discarding-message-because-the-queue-is-full/

**Reply by sebastianfer** [reply]
I did not move the robot yet since I thought the map would not be generated if that problem was not solved. Could you tell me where I can find the tf tree in the workspace and how can I change the parameters?

**Reply by justinIRBT** [reply]
The tf tree publishes on `/tf`.  In rviz2 you can add a viewer for TF that will show you the tree.  In the example, this line publishes the link between the robot tree and the lidar
https://github.com/iRobotEducation/create3_examples/blob/galactic/create3_lidar/launch/sensors_launch.py#L12

**Reply by sebastianfer** [reply]
So, I ran the command again today. I couldn't display the tf tree in rviz, but I did echo of the topic on Terminal. This is the last thing it returned: 

transforms:
- header:
    stamp:
      sec: 1658412365
      nanosec: 257626853
    frame_id: odom
  child_frame_id: base_link
  transform:
    translation:
      x: -2.061793088912964
      y: 3.644369125366211
      z: 0.08317188918590546
    rotation:
      x: -0.007724330760538578
      y: -0.000545671209692955
      z: 0.81966233253479
      w: 0.5727946758270264
- header:
    stamp:
      sec: 1658412365
      nanosec: 257626853
    frame_id: odom
  child_frame_id: base_footprint
  transform:
    translation:
      x: -2.0623789359229407
      y: 3.643689893423769
      z: 0.0
    rotation:
      x: 0.0
      y: 0.0
      z: 0.8196700811386108
      w: 0.5728358626365662
---

Does this mean it is doing the transformations properly?

**Comment by sebastianfer** [comment, marked as answer]
The solution to this problem can be found in #118. After fixing that, my Lidar was detected and registered.
